### PR TITLE
PLAT-1286 EmailToPdfConverter: Omit blank pages from PDF

### PR DESCRIPTION
**Pull request by alexander-schmidt**

### CODE REVIEW

File: platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/ImageToPdfConverterTest.java [hunk 1]
```diff
@@ -8,12 +8,10 @@
 
 public class ImageToPdfConverterTest extends Asserts {
 
-	private static final String MULTI_IMAGE_TIF = "multi-image.tif";
-
 	@Test
 	public void testConvertToPdf() throws IOException {
 
-		var pdfBytes = new ImageToPdfConverter(() -> getClass().getResourceAsStream(MULTI_IMAGE_TIF)).convertToPdf();
+		var pdfBytes = new ImageToPdfConverter(PdfTestFiles.TIF_MULTI_IMAGE.getResource()::getResourceAsStream).convertToPdf();
```
Comment (alexander-schmidt) [inline platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/ImageToPdfConverterTest.java:14]: Unrelated change: Moved this unsafely-referenced resource to `PdfTestFiles`.

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		} catch (IOException exception) {
+			throw new RuntimeException(exception);
+		}
+	}
+
+	private ArrayList<Integer> determineBlankPageIndexes(byte[] pdfBytes) {
+
+		try (var inputStream = new ByteArrayInputStream(pdfBytes)) {
+			var blankPageIndexes = new ArrayList<Integer>();
+			List<BufferedImage> pageImages = new PdfRenderer().setDpi(FAST_RENDERING_DPI).render(inputStream);
```
Comment (oliver-richers) [inline platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:50]: I wonder if `PdfRenderer` should also have an overload of `render(Supplier<PDDocument>)`. I don't know how expensive the `PDDocument.load` method is that we are calling here again.

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	 * @param pdfBytes
+	 *            the original PDF, as a byte array (never <i>null</i>)
+	 * @return the resulting PDF, as a byte array (never <i>null</i>)
+	 */
+	public byte[] removeLeadingBlankPages(byte[] pdfBytes) {
+
+		try (var document = PDDocument.load(pdfBytes)) {
+			determineBlankPageIndexes(pdfBytes)//
+				.stream()
+				.sorted(Collections.reverseOrder())
```
Comment (oliver-richers) [inline platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:36]: nice trick to reverse order here, but do we have a test with multiple blank pages interleaved with non-blank pages?

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+	private boolean isBlank(BufferedImage image) {
+
+		int height = image.getHeight();
+		int width = image.getWidth();
+
+		for (int x = 0; x < width; x++) {
+			for (int y = 0; y < height; y++) {
+				var color = new Color(image.getRGB(x, y));
+				if (color.getRed() < 255 || color.getGreen() < 255 || color.getBlue() < 255) {
```
Comment (oliver-richers) [inline platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:71]: - can't we just ask `if(color != WHITE)`? where `WHITE = 0xffffff`
- at least you could extract this into an `isWhite(int)` method, not every new developer knows what 255 means

File: platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/ImageToPdfConverterTest.java (hunk 1 above)
Comment (oliver-richers) [reply platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/ImageToPdfConverterTest.java:14]: ok

File: platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/PdfPageRemoverTest.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+		byte[] outputPdfBytes = new PdfPageRemover().removeLeadingBlankPages(inputPdfBytes);
+
+		try (var document = PDDocument.load(outputPdfBytes)) {
+			assertEquals(1, document.getNumberOfPages());
+			assertEquals("Third page\n", new PDFTextStripper().getText(document));
+		}
+	}
+
+	@Test
+	public void testRemoveLeadingBlankPagesWithoutLeadingBlankPages() throws IOException {
```
Comment (oliver-richers) [inline platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/PdfPageRemoverTest.java:26]: - the name `testRemoveLeadingBlankPagesWithoutLeadingBlankPages` is extremely confusing, you have to read the test to understand ;)
- suggest `testRemoveLeadingBlankPagesWithPdfWithNoLeadingBlankPages` or similar
- I am missing a test with multiple blank pages interleaved with non-blank pages to prevent regressions.
- also the name `PDF_WITHOUT_LEADING_BLANK_PAGES` does not indicate if there are any non-leading blank pages, is that intentional?

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java [hunk 6]
```diff
@@ -71,7 +72,8 @@ private byte[] convertToPdf(Part message) {
 
 		var buffer = new ByteArrayOutputStream();
 		convertToPdf(message, buffer);
-		return buffer.toByteArray();
+		byte[] pdfBytes = buffer.toByteArray();
+		return new PdfPageRemover().removeLeadingBlankPages(pdfBytes);
```
Comment (oliver-richers) [inline platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:76]:
```suggestion
		return new PdfPageRemover().removeLeadingBlankPages(buffer.toByteArray());
```

**Review by oliver-richers — CHANGES_REQUESTED** [review]

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+	/**
+	 * Identifies all consecutive leading blank pages in the given PDF byte
+	 * array, and returns a new PDF byte array in which those leading blank
+	 * pages are omitted.
+	 *
+	 * @param pdfBytes
+	 *            the original PDF, as a byte array (never <i>null</i>)
+	 * @return the resulting PDF, as a byte array (never <i>null</i>)
+	 */
+	public byte[] removeLeadingBlankPages(byte[] pdfBytes) {
```
[the remark this answers is not in the document]
Comment (alexander-schmidt) [reply platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:31]: makes sense

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java (hunk 2 above)
Comment (alexander-schmidt) [reply platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:50]: added an overload but opted for a non-supplier approach.

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java (hunk 3 above)
Comment (alexander-schmidt) [reply platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:36]: This is unrelated to interleaving. The following happens when remoding pages with either asending or descending indexes:
```
Remove with ascending indexes:
Available pages:  0 1 2 3 4 5 6
Pages to remove:  0 1 2
Result (wrong):   1 3 5 6

Remove with descending indexes:
Available pages:  0 1 2 3 4 5 6
Pages to remove:  2 1 0
Result (correct): 3 4 5 6
```

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java (hunk 4 above)
Comment (alexander-schmidt) [reply platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:71]: Now using `Color.WHITE`. I originally intended a fuzzy approach that considers slightly greyish values as blank - but I discarded that as unnecessary later-on. So this was an artifact.

File: platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java (hunk 6 above)
Comment (alexander-schmidt) [reply platform-core-module/src/main/java/com/softicar/platform/core/module/email/converter/EmailToPdfConverter.java:76]: inlined

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java (hunk 3 above)
Comment (oliver-richers) [reply platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfPageRemover.java:78]: I forgot that you only aim to remove leading pages, so all fine. Still, also with non-leading blank pages this would work.

File: platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfRenderer.java [hunk 8]
```diff
@@ -29,6 +28,9 @@ public class PdfRenderer {
 	private ImageType imageType;
 	private int maxRenderingThreads;
 
+	/**
+	 * Constructs a {@link PdfRenderer}.
```
Comment (oliver-richers) [inline platform-common-pdf/src/main/java/com/softicar/platform/common/pdf/PdfRenderer.java:32]: That is obvious. :D

File: platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/PdfPageRemoverTest.java (hunk 5 above)
Comment (alexander-schmidt) [reply platform-common-pdf/src/test/java/com/softicar/platform/common/pdf/PdfPageRemoverTest.java]: All of this was reworked to properly reflect that intermediate blank pages are removed as well. Pls take a look.

**Review by oliver-richers — APPROVED** [review]
